hail/scripts/benchmark_in_batch.py

```diff
@@ -0,0 +1,293 @@
+    deps: List[Job],  # dont reformat
+) -> Job:
+    j = b.new_job(name=f'{path}::{benchmark_name}-{trial}')
+    j.depends_on(*deps)
+    j.always_copy_output()
+
+    for varname, val in env.items():
+        if val is not None:
+            j.env(varname, val)
+
+    j.env('BENCHMARK_TRIAL_ID', str(trial))
+
+    j.command('mkdir -p benchmark-resources')
+    j.command(f'touch {j.ofile}')
+    j.command(f"""\
+python3 -m pytest {path} \
+    -Werror:::hail -Werror:::hailtop -Werror::ResourceWarning \
+    --log-cli-level=INFO \
+    -s \
+    -r A \
+    -vv \
+    --instafail \
+    --durations=50 \
+    --timeout=1800 \
+    --override-ini=tmp_path_retention_count=0 \
+    --override-ini=tmp_path_retention_policy=failed \
+    --output={j.ofile} \
+    --data-dir=benchmark-resources \
+    -k {benchmark_name}
+""")
+    return j.ofile
+
+
+def read_jsonl(p: Path):
+    logging.info(f'reading json lines from {p}.')
+    with p.open(encoding='utf-8') as r:
+        for line in r:
+            if len(line) > 1:
+                yield json.loads(line)
+
+
+def combine(output: Resource, files: List[Resource]) -> None:
+    init_logging()
+    n_files = len(files)
+    if n_files < 1:
+        raise ValueError("'combine' requires at least 1 file to merge")
+
+    logging.info(f'combining {len(files)} files')
+
+    jsonl = [line for f in files if (p := Path(f)).exists() for line in read_jsonl(p)]
+    jsonl.sort(key=lambda r: (r['path'], r['name'], r['trial']))
+
+    logging.info(f'Writing combine output to {output}')
+    logging.info(f'collected {len(jsonl)} benchmark jobs.')
+    with open(output, encoding='utf-8', mode='w+') as out:
+        for line in jsonl:
+            json.dump(line, out)
+            out.write('\n')
+
+
+def run_combine(args: Namespace) -> None:
+    combine(args.output, args.files)
+
+
+def make_combine_job(b: Batch, context: str, files: List[Resource]) -> Resource:
+    j = b.new_python_job(f'combine_output_{context}')
+    j.call(combine, j.ofile, files)
+    j.always_run()
+    return j.ofile
+
+
+def combine_results_as_tree(b: Batch, branch_factor: int, results: List[Resource]) -> Resource:
+    phase_i = 1
+
+    while len(results) > 1:
+        job_id = 0
+        results = [
+            (make_combine_job(b, f'phase{phase_i}_job{job_id}', files) if len(files) > 1 else files[0])
+            for files in chunk(branch_factor, results)
+            if (job_id := job_id + 1)
+        ]
+        phase_i += 1
+
+    return results[0]
+
+
+def run_submit(args: Namespace) -> None:
+    timestamp = time.strftime('%Y-%m-%d')
+    labelled_sha = args.sha + (f'-{args.label}' if args.label is not None else '')
+    object_prefix = getattr(args, 'object-prefix')
+    output_file = P.join(object_prefix, f'{timestamp}-{labelled_sha}.jsonl')
+
+    all_benchmarks = [
+        (path, name, trial)
+        for path, name, num_jobs in list_benchmarks(args.include, args.exclude)
+        for trial in range(num_jobs)
```

Comment:
note for the future: we haven't exposed job groups yet in the `hailtop.batch` API, but it could be nice to make a job group per benchmark that has multiple jobs associated with it. It's not immediately useful until we build a UI around it, but something to think about later on.

Reply:
Not sure if job groups are what we're after here given the randomisation of tests.